src/app/plugins/connectfour.plugin.ts

```diff
@@ -96,112 +100,214 @@ class ConnectFourGame {
+      (move) => Math.abs(bestValue - move.val) < this._getNormalizedDifficulty()
+    );
+    const randomMoveCol = moveOptions[Math.floor(Math.random() * moveOptions.length)].col;
+    this._dropPiece(randomMoveCol);
+  }
+
+  // We want the highest difficulty to allow for the least error.
+  private _getNormalizedDifficulty() {
+    return 1 / this._aiDifficulty;
+  }
+
+  // Evaluate the strength of the current board state,
+  // as it relates to Player 2's success.
+  private _evaluate(currentPlayer: number, depth: number) {
+    // If we have reached a win state, then the LAST move won.
+    if (this._checkWin()) {
+      return -4 * currentPlayer;
+    }
+    if (this._checkTie()) {
+      return 0;
+    }
+    // If we reached depth, then evaluate the board.
+    // We use a very simple evaluation: -(longest player one chain)
+    // This causes Lion to simply try to minimize Player 1's success,
+    // While our _checkWin case takes over when it's time to clinch victory.
+    if (depth === this._aiDepth) {
+      return this._longestChainOnBoard(1) - this._longestChainOnBoard(-1);
+    }
+
+    const moves: number[] = [];
+    for (let col = 0; col < this._cols; col++) {
+      if (!this._dropPiece(col, currentPlayer)) {
+        continue;
+      }
+
+      moves.push(this._evaluate(this.currentPlayer * -1, depth + 1));
+      this._removeTopPiece(col);
+    }
+
+    return this._arraySum(moves) / moves.length;
+  }
+
+  private _arraySum(array: number[]): number {
+    return array.reduce((acc, val) => (acc += val));
+  }
+
+  private _dropPiece(col: number, currentPlayer?: number): boolean {
+    let row = this._getFirstObjectInColumn(col);
+    // Place object one row above first object.
     row--;
 
     if (row === -1) {
       return false;
     }
 
-    this._board[row][col] = this.currentPlayer;
+    this._board[row][col] = currentPlayer ? currentPlayer : this.currentPlayer;
 
     return true;
   }
 
-  public checkWin(): boolean {
-    // Returns if a postion is valid and owned by the current player.
-    const validAndOwned: (row: number, col: number) => boolean = (row, col) => {
-      return (
-        row >= 0 &&
-        row < this._rows &&
-        col >= 0 &&
-        col < this._cols &&
-        this._board[row][col] === this.currentPlayer
-      );
-    };
+  private _removeTopPiece(col: number): void {
+    const row = this._getFirstObjectInColumn(col);
+
+    if (row === this._rows) {
+      return;
+    }
+
+    this._board[row][col] = 0;
+  }
+
+  private _getFirstObjectInColumn(col: number): number {
+    // Find first obstruction from top down in column.
+    let row = 0;
+    while (row < this._rows && this._board[row][col] === 0) {
+      row++;
+    }
+    return row;
+  }
+
+  private _longestChainOnBoard(currentPlayer?: number): number {
+    let longestChain = 0;
+    this._board.forEach((rowObj, rowIdx) => {
+      rowObj.forEach((_, colIdx) => {
+        longestChain = Math.max(
+          longestChain,
```

Comment:
use `reduce` so that you can do

```js
const longestchain = this._board.reduce(...)
```

Reply:
Better yet
```ts
return this._board.reduce(...)
```

Reply:
reduce will always return a value that is one dimension less than the object passed into it. How can I get around that for this case?

Reply:
As reference, this is what does not currently work.
`  private _longestChainOnBoard(currentPlayer?: number): number {
    return this._board.reduce((longestChain, rowObj, row) => {
      rowObj.forEach((_, col) => {
        longestChain = Math.max(
          longestChain,
          this._longestChainAtLocation(row, col, currentPlayer);
        );
      });
      return this._longestChainOnBoard;
    }, 0);`

Reply:
> reduce will always return a value that is one dimension less than the object passed into it. How can I get around that for this case?

 Not necessarily, you can make it return whatever you want

Reply:
> As reference, this is what does not currently work.
> 
> ```javascript
> 
> private _longestChainOnBoard(currentPlayer?: number): number {
> 
>     return this._board.reduce((longestChain, rowObj, row) => {
> 
>       rowObj.forEach((_, col) => {
> 
>         longestChain = Math.max(
> 
>           longestChain,
> 
>           this._longestChainAtLocation(row, col, currentPlayer);
> 
>         );
> 
>       });
> 
>       return this._longestChainOnBoard;
> 
>     }, 0);```

Foreach, doesn't return anything, you need a double reduce, or take the max of all 3 rows, so like
```ts 
return Math.max(this._board.map(row=> row.reduce((acc, rowObj, row) => Math.max(calculate the max, acc)
));
```

Reply:
Makes sense, reduce will always just feel useless to me I suppose ahahah